src/Rules/Library.hs

```diff
@@ -10,17 +13,53 @@ import Flavour
 import GHC
 import Oracles.ModuleFiles
 import Oracles.PackageData
+import Oracles.Dependencies (contextDependencies)
 import Settings
 import Settings.Path
 import Target
 import UserSettings
 import Util
 
+buildDynamicLib :: Context -> Rules ()
+buildDynamicLib context@Context{..} = do
+    -- macOS
+    matchGhcVersionedFilePath libPrefix "dylib" ?> buildDynamicLibUNIX
+    -- Linux
+    matchGhcVersionedFilePath libPrefix "so" ?> buildDynamicLibUNIX
+    -- TODO: Windows
+  where
+    path       = buildPath context
+    libPrefix  = path -/- "libHS" ++ pkgNameString package
+
+    buildDynamicLibUNIX so = do
+        deps <- contextDependencies context
+
+        forM_ deps $ \dep -> do
+            lib <- pkgLibraryFile dep
+            need [lib]
+
+        removeFile so
+        hsObjs   <- hsObjects    context
+        noHsObjs <- nonHsObjects context
+
+        -- This will create split objects if required (we don't track them
+        -- explicitly as this would needlessly bloat the Shake database).
+        need $ noHsObjs ++ hsObjs
```

Comment:
Yes there is code duplication since `buildDynamicLib` is adapted from `buildPackageLibrary`. Before that I am not sure if I should refactor or not, but let's refactor it.